keytransform/txndatastore.go

```diff
@@ -0,0 +1,415 @@
+package keytransform
+
+import (
+	"context"
+
+	ds "github.com/ipfs/go-datastore"
+	dsq "github.com/ipfs/go-datastore/query"
+)
+
+// WrapTxnDatastore wraps a given datastore with a KeyTransform function.
+// The resulting wrapped datastore will use the transform on all TxnDatastore
+// operations.
+func WrapTxnDatastore(child ds.TxnDatastore, t KeyTransform) *TxnDatastore {
+	if t == nil {
+		panic("t (KeyTransform) is nil")
+	}
+
+	if child == nil {
+		panic("child (ds.TxnDatastore) is nil")
+	}
+
+	return &TxnDatastore{child: child, KeyTransform: t}
+}
+
+// TxnDatastore keeps a KeyTransform function
+type TxnDatastore struct {
+	child ds.TxnDatastore
+
+	KeyTransform
+}
+
+var _ ds.Datastore = (*TxnDatastore)(nil)
+var _ ds.Batching = (*TxnDatastore)(nil)
+var _ ds.Shim = (*TxnDatastore)(nil)
+var _ ds.PersistentDatastore = (*TxnDatastore)(nil)
+var _ ds.CheckedDatastore = (*TxnDatastore)(nil)
+var _ ds.ScrubbedDatastore = (*TxnDatastore)(nil)
+var _ ds.GCDatastore = (*TxnDatastore)(nil)
+var _ ds.TxnDatastore = (*TxnDatastore)(nil)
+
+// Children implements ds.Shim
+func (d *TxnDatastore) Children() []ds.Datastore {
+	return []ds.Datastore{d.child}
+}
+
+// Put stores the given value, transforming the key first.
+func (d *TxnDatastore) Put(ctx context.Context, key ds.Key, value []byte) (err error) {
+	return d.child.Put(ctx, d.ConvertKey(key), value)
+}
```

Comment:
At first sight, re-implementing every method is not necessary.

The embedded datastore will provide them, that is why it is embedded right?. You just need to add NewTransaction() to this type.

Reply:
`Query()` and `prepareQuery()` need to be re-implemented to wrap things with the `KeyTransform`'s `t.InvertKey` and `t.ConvertKey`.  I believe everything else (except `NewTransaction()`) is already falling through to the child `ds.TxnDatastore`.

Reply:
Ah I see I think we can make the embedded child datastore a `keytransform.Datastore` and then all methods except `NewTransaction` can fall through to that. Will try this, thanks!

Reply:
This works! In this manner we can support every method except `NewTransaction` on the `TxnDatastore` by falling through to an embedded `*keytransform.Datastore` or `ds.TxnDatastore`. We still need to implement `Query` and `prepareQuery` on the `txnWrapper` in order to wrap the embedded `ds.Txn` with the `KeyTransform`.